### PR TITLE
refactor(increments-service): update routes and handlers for page views functionality

### PR DESCRIPTION
**Pull request by Yedidyar**
- Change route prefix from "/increments" to "/page-views" in the service and tests
- Implement new handler for batch increment requests with "/multi" endpoint
- Update existing increment logic to support multiple page views in a single request
- Adjust integration tests to reflect new endpoint structure and functionality

## Summary by CodeRabbit

* **New Features**
  * Added a new endpoint to support batch creation and updating of multiple page views in a single request.
* **Bug Fixes**
  * Improved error handling for invalid timestamp formats in batch operations.
* **Refactor**
  * All date and time handling for page views now use UTC for consistency.
  * The route prefix for page view endpoints has changed from "/increments" to "/page-views".
* **Tests**
  * Added integration tests for the new batch page views endpoint, including edge cases and aggregation logic.

### CODE REVIEW

**Comment by Yedidyar** [comment]
@coderabbitai review

**Comment by Yedidyar** [comment]
@coderabbitai give me places that I should improve the code

File: analytics-project/increments-service/handlers/index.ts [hunk 1]
```diff
@@ -13,6 +17,7 @@ export async function createOrUpdateSingleIncrementHandler(
     const { page, timestamp } = request.body;
 
     await request.server.incrementsService.incrementPage(page, timestamp);
+    return reply.status(200).send({ success: true });
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/handlers/index.ts]:
```suggestion
    return reply.status(200).send();
```

File: analytics-project/increments-service/handlers/index.ts [hunk 2]
```diff
@@ -26,6 +31,28 @@ export async function createOrUpdateSingleIncrementHandler(
...
 
+export async function createOrUpdateMultiIncrementHandler(
+  request: CreateOrUpdateMultiIncrementRequest,
+  reply: FastifyReply,
+) {
+  try {
+    const pageData = request.body;
+
+    await request.server.incrementsService.incrementMultiplePages(pageData);
+    return reply.status(200).send({ success: true });
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/handlers/index.ts]:
```suggestion
    return reply.status(200).send();
```

File: analytics-project/increments-service/handlers/index.ts [hunk 3]
```diff
@@ -26,6 +31,28 @@ export async function createOrUpdateSingleIncrementHandler(
...
 }
 
+export async function createOrUpdateMultiIncrementHandler(
+  request: CreateOrUpdateMultiIncrementRequest,
+  reply: FastifyReply,
+) {
+  try {
+    const pageData = request.body;
+
+    await request.server.incrementsService.incrementMultiplePages(pageData);
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/handlers/index.ts]:
```suggestion
    await request.server.incrementsService.incrementMultiplePages(request.body);
```

File: analytics-project/increments-service/integration.test.ts [hunk 4]
```diff
@@ -87,4 +87,149 @@ describe("Increments Service API Integration Tests", () => {
...
+        [altmanPage],
+      );
+
+      expect(altmanResult.rows.length).toBe(3);
+      expect(altmanResult.rows[0].views_count).toBe("103");
+      expect(altmanResult.rows[0].hour).toBe(0);
+      expect(altmanResult.rows[1].views_count).toBe("200");
+      expect(altmanResult.rows[1].hour).toBe(1);
+      expect(altmanResult.rows[2].views_count).toBe("405");
+      expect(altmanResult.rows[2].hour).toBe(2);
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/integration.test.ts]: Let's compare objects, checking it like this is tedious and ugly

File: analytics-project/increments-service/integration.test.ts [hunk 5]
```diff
@@ -87,4 +87,149 @@ describe("Increments Service API Integration Tests", () => {
...
+        "SELECT * FROM page_views WHERE name = $1 ORDER BY hour",
+        [muskPage],
+      );
+
+      expect(muskResult.rows.length).toBe(2);
+      expect(muskResult.rows[0].views_count).toBe("838");
+      expect(muskResult.rows[0].hour).toBe(0);
+      expect(muskResult.rows[1].views_count).toBe("654");
+      expect(muskResult.rows[1].hour).toBe(1);
+    });
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/integration.test.ts]: Same comment as above

File: analytics-project/increments-service/integration.test.ts [hunk 6]
```diff
@@ -87,4 +87,149 @@ describe("Increments Service API Integration Tests", () => {
...
+      );
+
+      // Should have 2 entries: one for hour 0 and one for hour 1
+      expect(result.rows.length).toBe(2);
+
+      // Hour 0 should have aggregated value: 100 + 200 + 300 = 600
+      expect(result.rows[0].hour).toBe(0);
+      expect(result.rows[0].views_count).toBe("600");
+
+      // Hour 1 should have the single value: 150
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/integration.test.ts]: I hate these comments that the AI adds to our code. Makes it looks less professional as they are unnecessary and add no value

File: analytics-project/increments-service/repositories/increments.ts [hunk 7]
```diff
@@ -20,17 +20,33 @@ export class IncrementsRepository {
     };
   }
 
+  async #executeIncrementQuery(
+    session: any,
```
[the remark this answers is not in the document]
Comment (tomer-shvadron) [reply analytics-project/increments-service/repositories/increments.ts]: Great catch but bad suggestion - `pg` exports necessary types

File: analytics-project/increments-service/repositories/increments.ts [hunk 8]
```diff
@@ -46,4 +62,137 @@ export class IncrementsRepository {
...
+        success: true,
+      }));
+
+      return {
+        successful: successfulResults,
+        failed: [],
+        totalProcessed: pageIncrements.length,
+        successCount: pageIncrements.length,
+        errorCount: 0,
+      };
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/repositories/increments.ts:128]: What do we think about atomicity in this section? What should we do if X our of Y rows fail?

File: analytics-project/increments-service/services/increments.service.ts [hunk 9]
```diff
@@ -5,9 +5,71 @@ export class IncrementsService {
 
   async incrementPage(page: string, timestamp: string) {
     const date = new Date(timestamp);
-    const hour = date.getHours();
+    const hour = date.getUTCHours();
     const value = 1;
 
-    await this.incrementsRepository.incrementPage(page, date, hour, value);
+    const dateOnly = date.toISOString().split("T")[0];
```
Comment (tomer-shvadron) [inline analytics-project/increments-service/services/increments.service.ts]: This should either be wrapped with a try-catch or we need to add validation over the date string that we get in the report

**Review by tomer-shvadron — APPROVED** [review]

File: analytics-project/increments-service/repositories/increments.ts (hunk 8 above)
Comment (Yedidyar) [reply analytics-project/increments-service/repositories/increments.ts:128]: In this case there is no partial failures

File: analytics-project/increments-service/services/increments.service.ts (hunk 9 above)
Comment (Yedidyar) [reply analytics-project/increments-service/services/increments.service.ts]: I don't understand this comment?
Why it should be wrapped with try catch?

File: analytics-project/increments-service/repositories/increments.ts (hunk 8 above)
Comment (tomer-shvadron) [reply analytics-project/increments-service/repositories/increments.ts:128]: Let's ask Ron about it

File: analytics-project/increments-service/services/increments.service.ts (hunk 9 above)
Comment (tomer-shvadron) [reply analytics-project/increments-service/services/increments.service.ts]: Thinking about it, we have no validation of the timestamp. That means that when we create the Date object from the timestamp string, we get an `Invalid Date` object, and then `getUTCHours` will cause an error.
So - validation of the timestamp (Regex, I assume) and wrapping with a try-catch to be on the safe side? WDYT?

**Review by tomer-shvadron — APPROVED** [review]

**Comment by Yedidyar** [comment]
> Thinking about it, we have no validation of the timestamp. That means that when we create the Date object from the timestamp string, we get an `Invalid Date` object, and then `getUTCHours` will cause an error.
> So - validation of the timestamp (Regex, I assume) and wrapping with a try-catch to be on the safe side? WDYT?

If We should add validation for all of the endpoints
@tomer-shvadron

**Review by tomer-shvadron — APPROVED** [review]
